### PR TITLE
4557/proxy vic admin

### LINKED ISSUE #4557: Make vicadmin respect proxy settings
Proxy settings are not applied to the applianceVM as a whole, so the settings must be read out of the environment (stored in an env var) and passed into the Golang HTTP client that does the health checks in vicadmin so that those health checks respect the proxy settings used elsewhere in the application.

### PR DESCRIPTION
**Pull request by morris-jason**
Fixes #4557 

- Removed HTTP[S]_PROXY env var for vicadmin. It was causing govmomi to attempt connections to vSphere over the proxy.
- WAN Checks now go over the proxy if it exists.

### CODE REVIEW

File: lib/vicadmin/validate.go [hunk 1]
```diff
@@ -172,20 +174,42 @@ func NewValidator(ctx context.Context, vch *config.VirtualContainerHostConfigSpe
...
+		proxy = sproxy
+	} else if hproxy := vchArgs["--http-proxy"]; hproxy != "" {
+		proxy = hproxy
+	}
+	if proxy != "" {
+		url, err := url.Parse(proxy)
 		if err != nil {
 			nwErrors = append(nwErrors, err)
 		} else {
-			conn.Close()
```
Comment (morris-jason) [inline lib/vicadmin/validate.go]: this is a diff bug

**Comment by sgairo** [comment]
Could you add tests please?

File: lib/vicadmin/validate.go [hunk 2]
```diff
@@ -172,20 +174,42 @@ func NewValidator(ctx context.Context, vch *config.VirtualContainerHostConfigSpe
...
+			client.Transport = &http.Transport{Proxy: http.ProxyURL(url)}
 		}
 	}
 
+	// preform the wan check
```
Comment (sgairo) [inline lib/vicadmin/validate.go]: "perform"

Comment (morris-jason) [inline lib/vicadmin/validate.go]: Oooh

File: lib/vicadmin/validate.go [hunk 3]
```diff
@@ -172,20 +174,42 @@ func NewValidator(ctx context.Context, vch *config.VirtualContainerHostConfigSpe
...
 	}
+
 	nwErrors := []error{}
 
-	for _, host := range hosts {
-		conn, err := net.DialTimeout("tcp", host, time.Minute/2)
+	// create a http client with a custom transport. The transport uses the [s/h]proxy defined in vicmachine create if applicable.
+	client := &http.Client{Timeout: 30 * time.Second}
+	vchArgs := make(map[string]string)
+	for _, args := range vch.VicMachineCreateOptions {
```
Comment (emlin) [inline lib/vicadmin/validate.go]: vicmachine create options is added for debug purpose so far. Why not read from env settings?

Comment (morris-jason) [inline lib/vicadmin/validate.go]: http.Client automatically picks up the proxy env var, which would be good for vicadmin expect govmomi also uses http.Client. When the proxy was supplied, govmomi was attempting to connect to vSphere/ESX over the proxy.

File: lib/vicadmin/validate.go [hunk 4]
```diff
@@ -172,20 +174,42 @@ func NewValidator(ctx context.Context, vch *config.VirtualContainerHostConfigSpe
...
+		splitArgs := strings.Split(args, "=")
+		vchArgs[splitArgs[0]] = splitArgs[1]
+	}
+	//priority given to https proxies
+	proxy := ""
+	if sproxy := vchArgs["--https-proxy"]; sproxy != "" {
```
Comment (anchal-agrawal) [inline lib/vicadmin/validate.go]:
```go
proxy := vchArgs["--https-proxy"]
if proxy == "" {
 	proxy = vchArgs["--http-proxy"]
}
```

The above snippet has the same behavior and is easier to read.

Comment (matthewavery) [inline lib/vicadmin/validate.go]: Additionally, we will likely want to check if the element is actually present in the map. It is true that even when an element is not present in a `map[string]string`the return value witll be `""`. However, it is still of interest of us whether or not the value was set to that or just defaulted to that value. 

```
 proxy, ok := vchArgs["--https-proxy"]
if !ok {
 	proxy = vchArgs["--http-proxy"]
}
```
when retrieving values from a map if `golang` you always get two values back. the value and a boolean for if the key was present.

File: lib/vicadmin/validate.go [hunk 5]
```diff
@@ -172,20 +174,42 @@ func NewValidator(ctx context.Context, vch *config.VirtualContainerHostConfigSpe
...
-	for _, host := range hosts {
-		conn, err := net.DialTimeout("tcp", host, time.Minute/2)
+	// create a http client with a custom transport. The transport uses the [s/h]proxy defined in vicmachine create if applicable.
+	client := &http.Client{Timeout: 30 * time.Second}
+	vchArgs := make(map[string]string)
+	for _, args := range vch.VicMachineCreateOptions {
+		splitArgs := strings.Split(args, "=")
+		vchArgs[splitArgs[0]] = splitArgs[1]
+	}
+	//priority given to https proxies
```
Comment (anchal-agrawal) [inline lib/vicadmin/validate.go]: Minor gofmt suggestion: have a space b/w `//` and the comment text.

File: lib/vicadmin/validate.go (hunk 4 above)
Comment (matthewavery) [inline lib/vicadmin/validate.go]: We also should likely have a constant for this raw string, just in case it gets reused in the future.

Comment (matthewavery) [inline lib/vicadmin/validate.go]: ditto for the `--http-proxy` string 😄

Comment (morris-jason) [inline lib/vicadmin/validate.go]: Using Anchal's code snippet, the [h/s]proxy will be available in the proxy var.

File: lib/vicadmin/validate.go (hunk 5 above)
Comment (morris-jason) [inline lib/vicadmin/validate.go]: Done, I should really pay attention to golinter.

File: lib/vicadmin/validate.go (hunk 4 above)
Comment (morris-jason) [inline lib/vicadmin/validate.go]: @matthewavery Seems like that just adds reading complexity in this case?

File: lib/vicadmin/validate.go (hunk 3 above)
Comment (emlin) [inline lib/vicadmin/validate.go]: hmm.. got your point.
So here in vicadmin, you'll need two connections, one to vSphere without proxy and one to googles needs proxy.
But you might not want to change the http proxy settings in guestinfo, which will involves upgrade change to support old version's VCH.
One suggestion is to look at the session creation code, to make it ignore default http proxy if is specified.

Comment (morris-jason) [inline lib/vicadmin/validate.go]: I don't think what I changes should alter guestinfo, just access it. 
Even if I changed the Session Client to ignore the http variable there still may be other places that use http.Client in vicadmin or it's vendored dependencies. 
Basically, it seems like the proper approach to only include the proxy where needed instead of excluding it in all the place it's not needed.

File: tests/test-cases/Group9-VIC-Admin/9-01-VICAdmin-ShowHTML.md [hunk 6]
```diff
@@ -22,3 +22,68 @@ This test requires that a vSphere environment be running and available
 * VICadmin responds with a log file indicating VCH init has begun reaping processes
 * VICadmin responds with log file indicating docker personality service has started
 * VICadmin responds with a ZIP file containing at a minimum the vmware.log file from the new container
+
+#Unauthenticated Test Cases
```
Comment (anchal-agrawal) [inline tests/test-cases/Group9-VIC-Admin/9-01-VICAdmin-ShowHTML.md]: Nit: this doesn't render the markdown heading correctly. `# Unauthenticated Test Cases` is the correct format. Bonus points for updating all headings in this file 😃

Comment (morris-jason) [inline tests/test-cases/Group9-VIC-Admin/9-01-VICAdmin-ShowHTML.md]: Weird..... Markdown Linter lied to me! Will update after the other build finished.

File: lib/vicadmin/validate.go (hunk 4 above)
Comment (matthewavery) [inline lib/vicadmin/validate.go]: @morris-jason for the change to checking the `ok` portion it is subtle but it does add a difference. In one case we are saying "did the user of this map give us this key with the value `""`"  the case where we use `!ok` protects us from this distinction. We will now be able to tell if someone attempted to give us `""` or did not attempt to populate that field at all.

File: lib/vicadmin/validate.go [hunk 7]
```diff
@@ -172,20 +174,40 @@ func NewValidator(ctx context.Context, vch *config.VirtualContainerHostConfigSpe
...
-		conn, err := net.DialTimeout("tcp", host, time.Minute/2)
+	// create a http client with a custom transport. The transport uses the [s/h]proxy defined in vicmachine create if applicable.
+	client := &http.Client{Timeout: 30 * time.Second}
+	vchArgs := make(map[string]string)
+	for _, args := range vch.VicMachineCreateOptions {
+		splitArgs := strings.Split(args, "=")
+		vchArgs[splitArgs[0]] = splitArgs[1]
+	}
+	// priority given to https proxies
+	proxy := vchArgs["--https-proxy"]
```
Comment (matthewavery) [inline lib/vicadmin/validate.go]: the raw strings here and 2 lines below should still be made constants.

Comment (morris-jason) [inline lib/vicadmin/validate.go]: Done.

File: lib/vicadmin/validate.go [hunk 8]
```diff
@@ -65,6 +67,8 @@ const (
 	GoodStatus     = template.HTML(`<i class="icon-ok"></i>`)
 	BadStatus      = template.HTML(`<i class="icon-attention"></i>`)
 	DefaultVCHName = ` `
+	HTTPSProxy     = "--https-proxy"
```
Comment (anchal-agrawal) [inline lib/vicadmin/validate.go]: These should be unexported, i.e. `httpsProxy` and `httpProxy` unless they're being used outside this package. 

Actually, it would be nice to leave these exported and use them in https://github.com/vmware/vic/blob/master/cmd/vic-machine/create/create.go (lines 453 and 461) so we can consolidate these variables. Check with @emlin on this approach.

File: lib/install/management/appliance.go [hunk 9]
```diff
@@ -536,12 +536,6 @@ func (d *Dispatcher) createAppliance(conf *config.VirtualContainerHostConfigSpec
 		},
 		Dir: "/home/vicadmin",
 	}
-	if settings.HTTPProxy != nil {
```
Comment (hmahmood) [inline lib/install/management/appliance.go]: You could just rename the env vars so that golang http client doesn't pick them up and then use them for the connectivity check.

Comment (morris-jason) [inline lib/install/management/appliance.go]: Will change to this approach following @emlin's comment that VicMachineCreateOptions was added for debug purposes.

**Comment by emlin** [comment]
this solution can workaround the issue for vsphere connection, but it still involves upgrade and reconfigure change.
you need a upgrade plugin following this instruction https://github.com/vmware/vic/blob/master/lib/migration/plugins/plugins.md
and you'll need to update the reconfigure a little bit. But if you can finish this PR quickly, I'll do it for you, otherwise, will let you know where to change it.
basically, any change in persisted configuration, will introduce these two more issues.

File: lib/vicadmin/validate.go [hunk 10]
```diff
@@ -172,20 +175,35 @@ func NewValidator(ctx context.Context, vch *config.VirtualContainerHostConfigSpe
...
+		"http://google.com",
+		"https://docker.io",
 	}
+
 	nwErrors := []error{}
 
-	for _, host := range hosts {
-		conn, err := net.DialTimeout("tcp", host, time.Minute/2)
+	// create a http client with a custom transport using the proxy from env vars
+	client := &http.Client{Timeout: 30 * time.Second}
```
Comment (caglar10ur) [inline lib/vicadmin/validate.go]: THANK YOU :) That time.Minute/2 was triggering my OCD

Comment (morris-jason) [inline lib/vicadmin/validate.go]: Haha :)

File: lib/migration/plugins/plugin6/vicadminProxyVarRename.go [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+		return nil
+	}
+	mapData := data.(map[string]string)
+	oldStruct := &VCHConfig{}
+	result := extraconfig.Decode(extraconfig.MapSource(mapData), oldStruct)
+	if result == nil {
+		return &errors.DecodeError{}
+	}
+	//translate old proxy env var keys into to proxy env var keys
+	newEnvs := make([]string, 0)
```
Comment (hmahmood) [inline lib/migration/plugins/plugin6/vicadminProxyVarRename.go]:
```go
var newEnvs []string
```

**Comment by morris-jason** [comment]
CI Is failing lib/migration tests. Not sure why - my plugin has a unique number. Someone please take a peek for me. 😃

File: lib/migration/plugins/plugin6/vicadminProxyVarRename.go [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+	oldStruct := &VCHConfig{}
+	result := extraconfig.Decode(extraconfig.MapSource(mapData), oldStruct)
+	if result == nil {
+		return &errors.DecodeError{}
+	}
+	//translate old proxy env var keys into to proxy env var keys
+	var newEnvs []string
+	for _, envVar := range oldStruct.Sessions["vicadmin"].Cmd.Env {
+		if strings.Contains(envVar, oldHProxy) {
+			proxyVarValue := strings.Split(envVar, "=")[1]
```
Comment (emlin) [inline lib/migration/plugins/plugin6/vicadminProxyVarRename.go]: might be better to check the split string length, to avoid panic here.

File: lib/migration/plugins/plugin6/vicadminProxyVarRename.go [hunk 13]
```diff
@@ -0,0 +1,105 @@
...
+			newEnvs = append(newEnvs, fmt.Sprintf("%s=%s", newHProxy, proxyVarValue))
+		} else if strings.Contains(envVar, oldSProxy) {
+			proxyVarValue := strings.Split(envVar, "=")[1]
```
Comment (emlin) [inline lib/migration/plugins/plugin6/vicadminProxyVarRename.go]: ditto

File: lib/migration/plugins/plugin6/vicadminProxyVarRename.go [hunk 14]
```diff
@@ -0,0 +1,105 @@
...
+	}
+	mapData := data.(map[string]string)
+	oldStruct := &VCHConfig{}
+	result := extraconfig.Decode(extraconfig.MapSource(mapData), oldStruct)
+	if result == nil {
+		return &errors.DecodeError{}
+	}
+	//translate old proxy env var keys into to proxy env var keys
+	var newEnvs []string
+	for _, envVar := range oldStruct.Sessions["vicadmin"].Cmd.Env {
```
Comment (emlin) [inline lib/migration/plugins/plugin6/vicadminProxyVarRename.go]: a constant for "vicadmin"?

**Comment by emlin** [comment]
You can try to update this line https://github.com/vmware/vic/blob/master/lib/migration/migrator_test.go#L35 to -1, because you have a plugin registered with latest version.

File: lib/migration/plugins/plugin6/vicadminProxyVarRename.go (hunk 12 above)
Comment (morris-jason) [inline lib/migration/plugins/plugin6/vicadminProxyVarRename.go]: Done.

File: lib/migration/plugins/plugin6/vicadminProxyVarRename.go (hunk 13 above)
Comment (morris-jason) [inline lib/migration/plugins/plugin6/vicadminProxyVarRename.go]: Done.

File: lib/migration/plugins/plugin6/vicadminProxyVarRename.go (hunk 14 above)
Comment (morris-jason) [inline lib/migration/plugins/plugin6/vicadminProxyVarRename.go]: Done.

File: lib/migration/plugins/init.go [hunk 15]
```diff
@@ -20,5 +20,5 @@ import (
 	_ "github.com/vmware/vic/lib/migration/plugins/plugin1"
 	_ "github.com/vmware/vic/lib/migration/plugins/plugin2"
 	_ "github.com/vmware/vic/lib/migration/plugins/plugin5"
-	_ "github.com/vmware/vic/lib/migration/plugins/plugin7"
+	_ "github.com/vmware/vic/lib/migration/plugins/plugin6"
```
Comment (emlin) [inline lib/migration/plugins/init.go]: why change it back to 6?

Comment (morris-jason) [inline lib/migration/plugins/init.go]: 7 was causing the test failure because of the order of versions in feature.go. I swapped it's place with plugin 6 and now it works.

File: lib/migration/feature/feature.go [hunk 16]
```diff
@@ -20,8 +20,8 @@ const (
 	TasksSupportedVersion
 	RenameSupportedVersion
 	MigrateRegistryVersion
-	ExecSupportedVersion
 	VicadminProxyVarRenameVersion
+	ExecSupportedVersion
```
Comment (emlin) [inline lib/migration/feature/feature.go]: this is wrong, you never insert your new plugin version before any others

File: lib/migration/plugins/init.go (hunk 15 above)
Comment (emlin) [inline lib/migration/plugins/init.go]: Did you try this?
`You can try to update this line https://github.com/vmware/vic/blob/master/lib/migration/migrator_test.go#L35 to -1, because you have a plugin registered with latest version.`

Comment (morris-jason) [inline lib/migration/plugins/init.go]: Yes. Still getting an error. The problem is that the migrator_test needs 2 plugin versions free to update. If my plugin is last there's only one plugin version free (the MaxPluginVersion).

Comment (morris-jason) [inline lib/migration/plugins/init.go]: Done.

File: lib/migration/feature/feature.go (hunk 16 above)
Comment (morris-jason) [inline lib/migration/feature/feature.go]: Done.

**Comment by anchal-agrawal** [comment]
@morris-jason please remember to add a descriptive commit message while merging the PR - "4557/proxy vic admin" doesn't convey the intent clearly.

File: lib/migration/plugins/plugin7/vicadminProxyVarRename.go [hunk 17]
```diff
@@ -0,0 +1,108 @@
...
+type Cmd struct {
+	Env []string `vic:"0.1" scope:"read-only" key:"Env"`
+}
+
+func (p *VicadminProxyVarRename) Migrate(ctx context.Context, s *session.Session, data interface{}) error {
+	defer trace.End(trace.Begin(fmt.Sprintf("%d", feature.VicadminProxyVarRenameVersion)))
+	if data == nil {
+		return nil
+	}
+	mapData := data.(map[string]string)
```
Comment (jzt) [inline lib/migration/plugins/plugin7/vicadminProxyVarRename.go]: Any chance `data` would be something other than a `map[string]string`?

File: lib/migration/plugins/plugin7/vicadminProxyVarRename.go [hunk 18]
```diff
@@ -0,0 +1,108 @@
...
+	oldStruct := &VCHConfig{}
+	result := extraconfig.Decode(extraconfig.MapSource(mapData), oldStruct)
+	if result == nil {
+		return &errors.DecodeError{}
+	}
+	//translate old proxy env var keys into to proxy env var keys
+
```
Comment (jzt) [inline lib/migration/plugins/plugin7/vicadminProxyVarRename.go]: whitespace looks like it should be on line 80 instead :)

File: lib/migration/plugins/plugin7/vicadminProxyVarRename.go [hunk 19]
```diff
@@ -0,0 +1,108 @@
...
+	if oldStruct.Sessions != nil && oldStruct.Sessions[sessionName] != nil && oldStruct.Sessions[sessionName].Cmd.Env != nil {
```
Comment (gigawhitlocks) [inline lib/migration/plugins/plugin7/vicadminProxyVarRename.go]: presumably you need to bail and return an error here if one of these is not nil.
In Go it is idiomatic to have the happy path unindented, so you likely want to refactor this to say, roughly
``` if oldStruct.Sessions == nil || oldStruct.Sessions[... {
          return error 
} 
// do stuff that was previously inside the if statement, but now on the outer level
```

Comment (morris-jason) [inline lib/migration/plugins/plugin7/vicadminProxyVarRename.go]: We can't error though because the plugin will be run regardless if it actually needs to. 
If I error, the migration process will stop whenever the proxies aren't defined and in all future versions. I'll refactor the inside code to be unindented and add a log statement though.

File: lib/migration/plugins/plugin7/vicadminProxyVarRename.go [hunk 20]
```diff
@@ -77,9 +77,12 @@ func (p *VicadminProxyVarRename) Migrate(ctx context.Context, s *session.Session
 	if result == nil {
 		return &errors.DecodeError{}
 	}
-	//translate old proxy env var keys into to proxy env var keys
 
-	if oldStruct.Sessions != nil && oldStruct.Sessions[sessionName] != nil && oldStruct.Sessions[sessionName].Cmd.Env != nil {
+	// translate old proxy env var keys into to proxy env var keys
+	// skip upgrading if the proxy isn't defined or something is wrong with the the vicadmin executor
+	if oldStruct.Sessions == nil || oldStruct.Sessions[sessionName] == nil || oldStruct.Sessions[sessionName].Cmd.Env == nil {
+		log.Debugln("vicadmin Session found. skipping proxy rename")
```
Comment (emlin) [inline lib/migration/plugins/plugin7/vicadminProxyVarRename.go]: not found

File: lib/migration/plugins/init.go [hunk 21]
```diff
@@ -20,4 +20,5 @@ import (
 	_ "github.com/vmware/vic/lib/migration/plugins/plugin1"
 	_ "github.com/vmware/vic/lib/migration/plugins/plugin2"
 	_ "github.com/vmware/vic/lib/migration/plugins/plugin5"
+	_ "github.com/vmware/vic/lib/migration/plugins/plugin7"
```
Comment (caglar10ur) [inline lib/migration/plugins/init.go]: we should send a pr to rename those, pluginX is really confusing. It could be the name of the enum for example, _ "github.com/vmware/vic/lib/migration/plugins/VicadminProxyVarRenameVersion"

File: lib/migration/plugins/plugin7/vicadminProxyVarRename.go (hunk 17 above)
Comment (emlin) [inline lib/migration/plugins/plugin7/vicadminProxyVarRename.go]: not likely. this is in the data migration chain, from a specific version to another one.
Even if we changed the data format in the future, the format for this specific plugin should still be this map only.
the input is defined as interface, just to make sure we don't break the data migration chain if meta data is moved somewhere else with different data format

File: lib/migration/plugins/plugin7/vicadminProxyVarRename.go (hunk 19 above)
Comment (gigawhitlocks) [inline lib/migration/plugins/plugin7/vicadminProxyVarRename.go]: OK cool so it's not an error condition, just looks like one. For what it's worth I've yet to write one of these migration plugins so you probably know more than me about them by now 😆
